### PR TITLE
`exists?` performance, primary key, response, tests

### PR DESCRIPTION
**Pull request by subvertallchris**
* Look for presence of a single value instead of counting all matches
* Allow String arguments, which will check for primary_key
* Fix response, only return boolean, no more `nil`

#1338

### CODE REVIEW

**Comment by subvertallchris** [comment]
Pings:
@cheerfulstoic 
@ProGM

File: lib/neo4j/active_node/query/query_proxy_methods.rb [hunk 1]
```diff
@@ -94,12 +94,13 @@ def include?(other, target = nil)
...
-          unless node_condition.is_a?(Integer) || node_condition.is_a?(Hash) || node_condition.nil?
-            fail(Neo4j::InvalidParameterError, ':exists? only accepts neo_ids')
+          unless [Integer, Fixnum, String, Hash, NilClass].include?(node_condition.class)
+            fail(Neo4j::InvalidParameterError, ':exists? only accepts ids or conditions')
           end
           query_with_target(target) do |var|
             start_q = exists_query_start(node_condition, var)
-            start_q.query.reorder.return("COUNT(#{var}) AS count").first.count > 0
+            result = start_q.query.reorder.return("ID(#{var}) AS found_ids LIMIT 1").first
+            !!result && result.found_ids > 0
```
Comment (cheerfulstoic) [inline lib/neo4j/active_node/query/query_proxy_methods.rb]: Shouldn't `found_ids` be `found_id`?  We're returning one row with one column and then I think we should be checking to see if the size of `result` is > 0, right?

File: lib/neo4j/active_node/query_methods.rb [hunk 2]
```diff
@@ -2,12 +2,13 @@ module Neo4j
...
       def exists?(node_condition = nil)
-        unless node_condition.is_a?(Integer) || node_condition.is_a?(Hash) || node_condition.nil?
+        unless [Integer, Fixnum, String, Hash, NilClass].include?(node_condition.class)
           fail(Neo4j::InvalidParameterError, ':exists? only accepts ids or conditions')
         end
         query_start = exists_query_start(node_condition)
         start_q = query_start.respond_to?(:query_as) ? query_start.query_as(:n) : query_start
-        start_q.return('COUNT(n) AS count').first.count > 0
+        result = start_q.return('ID(n) AS found_ids LIMIT 1').first
+        !!result && result.found_ids > 0
```
Comment (cheerfulstoic) [inline lib/neo4j/active_node/query_methods.rb]: Same comment as above

**Comment by cheerfulstoic** [comment]
Ah, I get it now.  Couldn't the first ID be 0, though, since the ID is an offset which starts at the beginning of a file?

You could even probably get a boolean from Cypher to make the Ruby code a bit simpler.  Something like:

```ruby
!!start_q.limit(1).with('ID(n) AS found_id').return('count(found_id) > 0').first
```

**Comment by subvertallchris** [comment]
That was one of the reasons why it was a "funny thing to check for!" I'm leaving the Cypher as is but should be good to go momentarily.

**Comment by cheerfulstoic** [comment]
:+1:

**Comment by subvertallchris** [comment]
Oof, I thought this was merged ages ago. Is the 8.0 branch still the best target, @cheerfulstoic?

**Comment by cheerfulstoic** [comment]
Yeah, `8.0.x` is still the running branch for the latest gem releases.  Make sure to merge to `master` afterward and push (I didn't do that one time and somebody else made a PR to solve part of the same problem)

**Comment by cheerfulstoic** [comment]
Since this is on `8.0.x` could you put in a `CHANGELOG` entry (either as `Unreleased` or with a new version)?

**Comment by subvertallchris** [comment]
Already pushed that straight to the branch and released as 8.0.8!

**Comment by cheerfulstoic** [comment]
Ah, right, I suppose I should have actually checked the `CHANGELOG` ;) . Thanks

**Pull request merged**